server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java

```diff
@@ -19,39 +19,217 @@
 import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
 
+import javax.mail.Header;
 import javax.mail.MessagingException;
+import javax.mail.internet.MimeMessage;
 
 import org.apache.james.core.MailAddress;
+import org.apache.james.mime4j.dom.Message;
+import org.apache.james.mime4j.stream.MimeConfig;
+import org.apache.james.util.mime.MessageContentExtractor;
+import org.apache.james.util.mime.MessageContentExtractor.MessageContent;
+import org.apache.james.util.streams.Iterators;
 import org.apache.mailet.Mail;
+import org.apache.mailet.PerRecipientHeaders;
 
+import com.github.fge.lambdas.Throwing;
 import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Multimap;
 
 public class MailDto {
-    public static MailDto fromMail(Mail mail) throws MessagingException {
+    public static MailDto fromMail(Mail mail, Set<AdditionalField> additionalFields) throws MessagingException, InnaccessibleFieldException {
         return new MailDto(mail.getName(),
             Optional.ofNullable(mail.getSender()).map(MailAddress::asString),
             mail.getRecipients().stream().map(MailAddress::asString).collect(Guavate.toImmutableList()),
             Optional.ofNullable(mail.getErrorMessage()),
-            Optional.ofNullable(mail.getState()));
+            Optional.ofNullable(mail.getState()),
+            Optional.ofNullable(mail.getRemoteHost()),
+            Optional.ofNullable(mail.getRemoteAddr()),
+            Optional.ofNullable(mail.getLastUpdated()),
+            fetchAttributes(additionalFields, mail),
+            fetchPerRecipientsHeaders(additionalFields, mail),
+            fetchHeaders(additionalFields, mail),
+            fetchTextBody(additionalFields, mail),
+            fetchHtmlBody(additionalFields, mail),
+            fetchMessageSize(additionalFields, mail));
+    }
+
+    private static Optional<Long> fetchMessageSize(Set<AdditionalField> additionalFields, Mail mail) throws InnaccessibleFieldException {
+        if (!additionalFields.contains(AdditionalField.MESSAGE_SIZE)) {
+            return Optional.empty();
+        }
+        try {
+            return Optional.of(mail.getMessageSize());
+        } catch (MessagingException e) {
+            throw new InnaccessibleFieldException(AdditionalField.MESSAGE_SIZE, e);
+        }
+    }
+
+    private static Optional<String> fetchTextBody(Set<AdditionalField> additionalFields, Mail mail) throws InnaccessibleFieldException {
+        if (!additionalFields.contains(AdditionalField.TEXT_BODY)) {
+            return Optional.empty();
+        }
+
+        try {
+            MessageContentExtractor extractor = new MessageContentExtractor();
+            return Optional.ofNullable(mail.getMessage())
+                    .map(Throwing.function(MailDto::convertMessage).sneakyThrow())
+                    .map(Throwing.function(extractor::extract).sneakyThrow())
```

Comment:
It's a shame... MIME parsing is memory and time expensive, we should ensure we do parse that message only once, in `fromMail` then only use the MessageCOntent in `fetchHtmlBody` & `fetchTextBody`.